### DISCUSSION TITLE
Migrate manny thanos clusters into one with multi-tenancy

### DISCUSSION THREAD

**Discussion by Avigdorrr** [thanos-io/thanos#7580] [Questions & Answers]
Currently we have a lot thanos clusters (25 clusters) that each have a querier, receiver, store and an object storage bucket. We want to migrate to a big cluster with multi-tenancy.
Is there a way to migrate the old data in each bucket to a new bucket with blocks that will consist that data in a multi-tenancy format? 
It means it needs to edit the blocks to match the multi-tenancy way with a tsdb per each tenant

I currently have that data with the label key “env” for each cluster. so if it possible to order and group the new blocks by that label values and make that an external label, it seems like it will do the work

Any way that I can achieve that goal?

**Comment by MeNsaaH** [comment]
Hey @Avigdorrr , I'm curious how you eventually achieved this. This is something I looking at doing and looking for pointers. Thanks in advance!

**Reply by Avigdorrr** [reply]
Hey @MeNsaaH,

We actually ended up deciding not to go through with this migration in the end.

That being said, from the research we did back then, I think the way to do it would be manipulating the block metadata directly. I'm not 100% sure and this definitely needs to be tested, but you would essentially need to write a script to iterate through your old object storage buckets, download the meta.json file for every block, inject your desired tenant label into the thanos.labels section, and then upload the modified meta.json back.

Since Thanos relies on that metadata file to map blocks to specific tenants, modifying it retroactively should theoretically do the trick to get the historical data recognized in a multi-tenant setup. But again, definitely test it thoroughly on a small, non-production bucket first before touching your actual historical data.

**Comment by ezeqme** [comment]
@Avigdorrr Please, what were the problems that motivated you to make this architectural decision? I'm curious because I'm doing exactly the opposite - breaking down large Thanos clusters into smaller clusters with multi-tenant support.
What issues were you encountering? Performance problems, operational complexity, resource utilization inefficiencies, or something else entirely?

**Reply by Avigdorrr** [reply]
Hey @ezeqme,

For us, the main driver was mostly operational complexity. On top of that, in our specific case, we couldn't really control the incoming metric volume in a predictable way. Because of those unpredictable loads, resource allocation across so many individual clusters became a constantly moving target that was really hard to fix efficiently.

Ultimately, tenancy is a flexible mechanism. You can implement it on a single large cluster (soft tenancy), across many separate clusters (hard tenancy), or even use a combination of both approaches depending on your specific isolation and performance needs.

Hope that gives some helpful context on our thought process!
